### PR TITLE
chore(toolchain): bump to 2022-12-12

### LINKED ISSUE #6087: streaming: fix hash join degree table scan
```
             let mut table_data = vec![];
             let mut degree_table_data = vec![];

             #[for_await]
             for x in table_iter {
                 table_data.push(x?);
             }

             #[for_await]
             for x in degree_table_iter {
                 degree_table_data.push(x?);
             }
```

Now we fetch ALL the data before start processing, instead of fetch-while-processing.

Part of https://github.com/risingwavelabs/risingwave/issues/5977, can be fixed after the compiler bug is fixed.

### LINKED ISSUE #6835: compiler bug(?): daily release build takes 2 hours
https://buildkite.com/risingwavelabs/docker/builds/13182#0184ed58-e171-4d20-bc20-55d47cf2ce91
SCR-20221210-54

Normal on my laptop.

### LINKED ISSUE #6864: revert workarounds in deterministic scaling test

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Bump the toolchain to 2022-12-12, which should contain the fix to https://github.com/rust-lang/rust/issues/103423. We're now able to revert some workarounds, as explained in #6103.

This is mainly to fix the excessively slow compilation speed under release profile, like https://buildkite.com/risingwavelabs/docker/builds/13181.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #6103
- Close #6835 
- Close #6087
- Close #6864

### CODE REVIEW

**Comment by BugenZhao** [comment]
@huangjw806 Could you please help to test whether building the docker image under release profile finishes in a reasonable time, under this branch?

**Review by huangjw806 — APPROVED** [review]
LGTM

**Comment by huangjw806** [comment]
> ou please help to test whether building the docker image under release profile finishes in a reasonable time, under this branch?

Sure.

**Review by xxchan — APPROVED** [review]

**Review by wangrunji0408 — APPROVED** [review]

**Comment by BugenZhao** [comment]
https://github.com/rust-lang/rust/issues/105009 🥵

**Comment by BugenZhao** [comment]
It seems like it's https://github.com/risingwavelabs/risingwave/pull/6613 that causes the slow compilation before bumping and ICE after bumping.

cc @wenym1 Would you please help to take a look and clean some suspicious sugars or magics in that PR? 🥵

**Comment by wenym1** [comment]
> It seems like it's https://github.com/risingwavelabs/risingwave/pull/6613 that causes the slow compilation before bumping and ICE after bumping.

cc @wenym1 Would you please help to take a look and clean some suspicious sugars or magics in that PR? 🥵

In #6613, we only wrap the previous `StateStoreIter` with a `try_stream`. I am not sure whether this will cause the slow compilation. Can you provide me a way to reproduce the problem locally?

**Comment by BugenZhao** [comment]
> Can you provide me a way to reproduce the problem locally?

I guess compiling with release profile can reproduce this, just like what we do when building image on CI.

https://github.com/risingwavelabs/risingwave/blob/353837d0029778c8ca06055fc069e8834f9fa65c/docker/Dockerfile#L35

**Comment by BugenZhao** [comment]
https://github.com/rust-lang/rust/issues/105009 has been fixed and I've successfully bumped the toolchain to 12-12. Seems that everything is working well.

cc @wenym1 There's no need to do anything with #6613 now. :)

File: src/storage/src/lib.rs [hunk 1]
```diff
@@ -41,7 +41,7 @@
 #![feature(error_generic_member_access)]
 #![feature(provide_any)]
 #![feature(let_chains)]
-#![expect(clippy::result_large_err, reason = "FIXME: HummockError is large")]
+// #![expect(clippy::result_large_err, reason = "FIXME: HummockError is large")]
```
Comment (BugenZhao) [inline src/storage/src/lib.rs:44]:
```suggestion
```

File: src/storage/src/table/streaming_table/mem_table.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ pub enum MemTableError {
     },
 }
 
-type Result<T> = std::result::Result<T, MemTableError>;
+type Result<T> = std::result::Result<T, Box<MemTableError>>;
```
Comment (BugenZhao) [inline src/storage/src/table/streaming_table/mem_table.rs:46]: This error is also large.

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 3]
```diff
@@ -354,34 +353,14 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
...
-            // with the same stream key in degree table.
-            let zipped_iter = zip_by_order_key(
-                futures::stream::iter(table_data.into_iter()),
-                futures::stream::iter(degree_table_data.into_iter()),
-            );
-
-            #[for_await]
-            for row_and_degree in zipped_iter {
-                let (row, degree): (Cow<'_, Row>, Cow<'_, Row>) = row_and_degree?;
+            for (row, degree) in table_iter.zip(degree_table_iter) {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:357]: cc @yuhao-su Am I correct here?

File: e2e_test/ch_benchmark/batch/q3.slt.part [hunk 4]
```diff
@@ -16,21 +16,21 @@ where 	 c_state like 'A%'
 group by ol_o_id, ol_w_id, ol_d_id, o_entry_d
 order by revenue desc, o_entry_d;
 ----
-    2111        1        4       21  2015-11-22 00:00:00
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply e2e_test/ch_benchmark/batch/q3.slt.part:19]: Oops. Will keep `rowsort` here.

File: e2e_test/ch_benchmark/batch/q10.slt.part [hunk 5]
```diff
@@ -10,26 +10,25 @@ where	 c_id = o_c_id
     	 and o_entry_d >= '2007-01-02 00:00:00.000000'
     	 and o_entry_d <= ol_delivery_d
     	 and n_nationkey = ascii(substr(c_state,1,1)) - 48
-group by c_id, c_last, c_city, c_phone, n_name
-order by revenue desc;
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply e2e_test/ch_benchmark/batch/q10.slt.part:14]: I think `order by` makes no sense under `rowsort`?

File: ci/Dockerfile [hunk 6]
```diff
@@ -28,7 +28,7 @@ RUN rustup component add rustfmt llvm-tools-preview clippy
 
 # install build tools
 RUN cargo install cargo-llvm-cov cargo-nextest cargo-udeps cargo-hakari cargo-sort cargo-make cargo-cache \
-    && cargo install --git https://github.com/risinglightdb/sqllogictest-rs --bin sqllogictest \
+    && cargo install --git https://github.com/risinglightdb/sqllogictest-rs --rev 65b122f --bin sqllogictest \
```
Comment (BugenZhao) [inline ci/Dockerfile:31]: I'd like to specify the revision to make the bump of sqllogictest more explicit. 😄 It's the lastest revision on `main` now. cc @xxchan

Comment (xxchan) [reply ci/Dockerfile:31]: very good

**Pull request merged**